Support boolean operators grammar in contains search condition

jaspal007 opened the pull request:
Previously in Babelfish the support of Boolean operators was not enabled. To fix this we have created a grammar to allow customers to create search expressions with boolean operators `AND`, `OR` and `AND NOT` along with complex expressions using parenthesis over `simple_term` and `prefix_term` operands

### 1. Issue:
 - Fulltext search using boolean operators in contains search condition is not supported in Babelfish. When a boolean operator is passed into the search string, it throws `Full-text search conditions with boolean operators are not currently supported in Babelfish` error
   - current output
     ```
     1> SELECT * FROM bool_table WHERE CONTAINS(col, 'one AND two')
     2> GO
     Msg 33557097, Level 16, State 1, Server persist-rskanwar-instance-bbf-17-2-test, Line 1
     Full-text search conditions with boolean operators are not currently supported in Babelfish
     ``` 

### 2. Changes made to fix the issue:
 - created a grammar for boolean operators to be supported in contains search condition
 - added the support for valid boolean operators: `AND`, `OR` and `AND NOT` 
 - added the support for different forms of boolean operators
    ```
    AND        and           &
    OR         or            |
    AND NOT    and not       &!
    ```
 - handled `NOT` boolean operator to throw syntax error as it is not an allowed boolean operator in SQL SERVER Full-Text Search
 - extended the contains search condition grammar to enable customers to create complex search strings with the help of parenthesis and boolean operators
 - added a custom function to merge the translated tokens into one single translated search string. 
 - customer can perform a full text search over a combination of `simple_term` and `prefix_term` operands with the help of boolean operators. 


TASK: BABEL-4382
Authored-by: Jaspal Singh [ijaspals@amazon.com](mailto:ijaspals@amazon.com)
Signed-off-by: Jaspal Singh [ijaspals@amazon.com](mailto:ijaspals@amazon.com)
### Issues Resolved

BABEL-4382

### Test Scenarios Covered ###
* **Use case based -** YES


* **Boundary conditions -** YES


* **Arbitrary inputs -**


* **Negative test cases -** YES


* **Minor version upgrade tests -** 


* **Major version upgrade tests -**


* **Performance tests -**


* **Tooling impact -**


* **Client tests -**



### Check List
- [x] Commits are signed per the DCO using --signoff 

By submitting this pull request, I confirm that my contribution is under the terms of the Apache 2.0 and PostgreSQL licenses, and grant any person obtaining a copy of the contribution permission to relicense all or a portion of my contribution to the PostgreSQL License solely to contribute all or a portion of my contribution to the PostgreSQL open source project.

For more information on following Developer Certificate of Origin and signing off your commits, please check [here](https://github.com/babelfish-for-postgresql/babelfish_extensions/blob/main/CONTRIBUTING.md#developer-certificate-of-origin).

Commit:
Support boolean operators grammar in contains search condition

Previously in Babelfish the support of Boolean operators was not enabled. To fix this we have created a grammar to allow customers to create search expressions with boolean operators 'AND', 'OR' and 'AND NOT' along with complex expressions using parenthesis over the simple_term and prefix_term operands

TASK: BABEL-4382
Signed-off-by: Jaspal Singh ijaspals@amazon.com

sumitj824 commented on contrib/babelfishpg_tsql/src/fts_scan.l:
```diff
@@ -21,23 +21,28 @@ static YY_BUFFER_STATE scanbufhandle;
-"OR"            { yylval = "OR"; yyerror(NULL, "Full-text search conditions with boolean operators are not currently supported in Babelfish"); return OR_TOKEN; }
-"NOT"			{ yylval = "NOT"; yyerror(NULL, "Full-text search conditions with boolean operators are not currently supported in Babelfish"); return NOT_TOKEN; }
+"AND"           { yylval = "AND"; return AND_TOKEN; }
+"and"           { yylval = "AND"; return AND_TOKEN; }
```
Can you please check if we really need to add operator for lowercase? If yes, then what will happen if the operator provided in input is in mixedcase.

sumitj824 commented on contrib/babelfishpg_tsql/src/fts_scan.l:
```diff
@@ -21,23 +21,28 @@ static YY_BUFFER_STATE scanbufhandle;
 "INFLECTIONAL"  { yylval = "INFLECTIONAL"; return INFLECTIONAL_TOKEN; }
 "THESAURUS"     { yylval = "THESAURUS"; return THESAURUS_TOKEN; }
-"FORMSOF"       { yylval = "FORMSOF"; return FORMSOF_TOKEN; }
+"FORMSOF"       { yylval = "FORMSOF"; yyerror(NULL, "Generation term is not currently supported in Babelfish"); return FORMSOF_TOKEN; }
```
Can you please explain that why did we make this change? Also, I think if you throwing error with yyerror then syntax error code will used instead we should explore we can have right error code.

sumitj824 commented on contrib/babelfishpg_tsql/src/fts_parser.y:
```diff
@@ -419,7 +512,7 @@ replaceMultipleSpacesAndSpecialChars(char* input, char **str1, char **str2, bool
             ereport(ERROR,
                 (errcode(ERRCODE_FEATURE_NOT_SUPPORTED),
-                 errmsg("Full-text search conditions with boolean operators are not currently supported in Babelfish")));
+                 errmsg("Syntax error in the full-text search condition")));
```
Please change the error code from `ERRCODE_FEATURE_NOT_SUPPORTED` to `ERRCODE_SYNTAX_ERROR`.

sumitj824 commented on contrib/babelfishpg_tsql/src/fts_parser.y:
```diff
@@ -22,17 +22,19 @@ extern char     *replace_special_chars_fts_impl(char *input_str);
 %}
 
-%token WORD_TOKEN WS_TOKEN TEXT_TOKEN PREFIX_TERM_TOKEN GENERATION_TERM_TOKEN AND_TOKEN NOT_TOKEN AND_NOT_TOKEN OR_TOKEN INFLECTIONAL_TOKEN THESAURUS_TOKEN FORMSOF_TOKEN O_PAREN_TOKEN C_PAREN_TOKEN COMMA_TOKEN SPECIAL_CHAR_TOKEN NON_ENGLISH_TOKEN
+%token WORD_TOKEN WS_TOKEN TEXT_TOKEN PREFIX_TERM_TOKEN GENERATION_TERM_TOKEN AND_TOKEN NOT_TOKEN AND_NOT_TOKEN OR_TOKEN O_PAREN_TOKEN C_PAREN_TOKEN INFLECTIONAL_TOKEN THESAURUS_TOKEN FORMSOF_TOKEN COMMA_TOKEN SPECIAL_CHAR_TOKEN NON_ENGLISH_TOKEN
```
Does ordering in which token names are listed really matters here?

sumitj824 replied:
Please remove `GENERATION_TERM_TOKEN` if we aren't using this anymore.

sumitj824 commented on contrib/babelfishpg_tsql/src/fts_parser.y:
```diff
@@ -43,58 +45,105 @@ static void     replaceMultipleSpacesAndSpecialChars(char* input, char **str1, c
+    ;
+
+
+multiple_term:
+    search_term {
+        $$ = $1;
+    }
+    | enclosed_term {
+        $$ = $1;
+    }
+    | search_term bool_operator multiple_term {
+        $$ = mergeTokens($1, $2, $3);
+    }
+    | enclosed_term bool_operator multiple_term {
+        $$ = mergeTokens($1, $2, $3);
+    }
+    ;
```
Suggestion: We can have this grammar something like this 
```
multiple_term:
    search_term opt_node
    | enclosed_term opt_node
    ;

opt_node:
    bool_operator multiple_term
    | /*EMPTY */
```

sumitj824 commented on contrib/babelfishpg_tsql/src/fts_parser.y:
```diff
@@ -43,58 +45,105 @@ static void     replaceMultipleSpacesAndSpecialChars(char* input, char **str1, c
+    ;
+
+
+enclosed_term:
```
Same as previous comment, this grammar can be simplified.

sumitj824 commented on contrib/babelfishpg_tsql/src/fts_parser.y:
```diff
@@ -43,58 +45,105 @@ static void     replaceMultipleSpacesAndSpecialChars(char* input, char **str1, c
+    }
+    ;
+
+search_term:
     generation_term
     | simple_term
     | prefix_term
     ;
+    
 
 simple_term:
-    WORD_TOKEN  {
-        *result = translate_simple_term($1);
-    }
-    | TEXT_TOKEN {
-        *result = translate_simple_term($1);
-    }
-    | WS_TOKEN WORD_TOKEN {
-        *result = translate_simple_term($2);
+    keyword {
+        $$ = $1;
     }
-    | WORD_TOKEN WS_TOKEN {
-        *result = translate_simple_term($1);
+    | WS_TOKEN keyword {
+        $$ = $2;
     }
-    | WS_TOKEN WORD_TOKEN WS_TOKEN {
-        *result = translate_simple_term($2);
+    | keyword WS_TOKEN {
+        $$ = $1;
     }
-    | WS_TOKEN TEXT_TOKEN {
-        *result = translate_simple_term($2);
+    | WS_TOKEN keyword WS_TOKEN {
+        $$ = $2;
     }
-    | TEXT_TOKEN WS_TOKEN {
-        *result = translate_simple_term($1);
+    ;
```
This can be also simplified to single rule.

sumitj824 commented on contrib/babelfishpg_tsql/src/fts_scan.l:
```diff
@@ -21,23 +21,28 @@ static YY_BUFFER_STATE scanbufhandle;
+"AND"           { yylval = "AND"; return AND_TOKEN; }
+"and"           { yylval = "AND"; return AND_TOKEN; }
+"&"           	{ yylval = "AND"; return AND_TOKEN; }
+"AND NOT"       { yylval = "AND NOT"; return AND_NOT_TOKEN; }
```
Can you please verify if this will also work when there are more than one space in between in AND and NOT?

sumitj824 commented on contrib/babelfishpg_tsql/src/fts_parser.y:
```diff
@@ -156,20 +224,42 @@ static char
+        hasSpecialChars = true;
+    }
+    
+    /*
```
NIT: Please follow the commenting style as per PG.

Commit:
Merge remote-tracking branch 'upstream/BABEL_5_X_DEV' into BABEL-4382-5_X

roshan0708 commented on contrib/babelfishpg_tsql/src/fts_parser.y:
```diff
@@ -156,20 +224,42 @@ static char
+
+    /*
+     * all the simple term search strings will be translated with
+     * a paranthesis around them as along with boolean expression 
```
nit: Parenthesis*

Commit:
Support boolean operators grammar in contains search condition

Previously in Babelfish the support of Boolean operators was not enabled. To fix this we have created a grammar to allow customers to create search expressions with boolean operators 'AND', 'OR' and 'AND NOT' along with complex expressions using parenthesis over the simple_term and prefix_term operands

TASK: BABEL-4382
Signed-off-by: Jaspal Singh ijaspals@amazon.com

Commit:
Support boolean operators grammar in contains search condition

Previously in Babelfish the support of Boolean operators was not enabled. To fix this we have created a grammar to allow customers to create search expressions with boolean operators 'AND', 'OR' and 'AND NOT' along with complex expressions using parenthesis over the simple_term and prefix_term operands.

TASK: BABEL-4382
Signed-off-by: Jaspal Singh ijaspals@amazon.com

Commit:
Merge remote-tracking branch 'upstream/BABEL_5_X_DEV' into BABEL-4382-5_X

roshan0708 commented on contrib/babelfishpg_tsql/src/fts_scan.l:
```diff
@@ -21,36 +25,46 @@ static YY_BUFFER_STATE scanbufhandle;
+"&!"       		{ yylval = "AND NOT"; return AND_NOT_TOKEN; }
+"OR"           		{ yylval = "OR"; return OR_TOKEN; }
+"|"            		{ yylval = "OR"; return OR_TOKEN; }
+"NOT"			{ yylval = "NOT"; yyerror(NULL, "Syntax error in the full-text search condition"); return NOT_TOKEN; }
```
Replace this with `yyerror_with_code`, use syntax error code. Check other occurrences of `yyerror` as well in `fts_parser.y` file

Commit:
Support boolean operators grammar in contains search condition

Previously in Babelfish the support of Boolean operators was not enabled. To fix this we have created a grammar to allow customers to create search expressions with boolean operators 'AND', 'OR' and 'AND NOT' along with complex expressions using parenthesis over the simple_term and prefix_term operands.

TASK: BABEL-4382
Signed-off-by: Jaspal Singh ijaspals@amazon.com

Commit:
Merge remote-tracking branch 'upstream/BABEL_5_X_DEV' into BABEL-4382-5_X

Commit:
Merge remote-tracking branch 'upstream/BABEL_5_X_DEV' into BABEL-4382-5_X

Commit:
Support boolean operators grammar in contains search condition

Previously in Babelfish the support of Boolean operators was not enabled. To fix this we have created a grammar to allow customers to create search expressions with boolean operators 'AND', 'OR' and 'AND NOT' along with complex expressions using parenthesis over the simple_term and prefix_term operands.

TASK: BABEL-4382
Signed-off-by: Jaspal Singh ijaspals@amazon.com

jaspal007 commented:
Added some more test cases

sumitj824 commented on contrib/babelfishpg_tsql/src/fts_parser.y:
```diff
@@ -43,68 +45,100 @@ static void     replaceMultipleSpacesAndSpecialChars(char* input, char **str1, c
-        *result = translate_simple_term($2);
+    ;
+
+opt_node:
```
NIT: please use meaningful name.

sumitj824 commented on contrib/babelfishpg_tsql/src/fts_parser.y:
```diff
@@ -117,8 +151,31 @@ simple_term_list:
     }
     ;
 
+white_space:
```
NIT: This should be named as opt_white_space.

sumitj824 commented on contrib/babelfishpg_tsql/src/fts_parser.y:
```diff
@@ -43,68 +45,100 @@ static void     replaceMultipleSpacesAndSpecialChars(char* input, char **str1, c
+    ;
+
+enclosed_term:
+    white_space O_PAREN_TOKEN multiple_term C_PAREN_TOKEN white_space {
```
Please check if multiple whitespaces are allowed or not. If allowed then please use `(WS_TOKEN)*` directly here. Please check at all the place where we are using such nodes.

sumitj824 commented on contrib/babelfishpg_tsql/src/fts_scan.l:
```diff
@@ -21,36 +22,47 @@ static YY_BUFFER_STATE scanbufhandle;
 \"[^"]+\*\"               																							{ if (yytext[strlen(yytext)-2] == ' ') { yyerror(NULL, "Syntax error, space is not allowed before *"); } yylval = yytext; return PREFIX_TERM_TOKEN; }  /* Handle prefix terms */
 \"[^"]*\"                																						    { yylval = yytext; return TEXT_TOKEN; }  /* Handle double-quoted phrases */
-[a-zA-Z0-9~`!@#$%^&*()+_\-={[}\]\|\\:;"'<,>.?/]+ 																	{ if (strncasecmp(yytext, "FORMSOF(", strlen("FORMSOF(")) == 0) { yyerror(NULL, "Generation term is not currently supported in Babelfish"); } yylval = yytext; return WORD_TOKEN; }  /* Handle individual words */
+[a-zA-Z0-9~`!@#$%^&*+_\-={[}\]|\\:;"'<,>.?/]+																	{ if (strncasecmp(yytext, "FORMSOF(", strlen("FORMSOF(")) == 0) { yyerror(NULL, "Generation term is not currently supported in Babelfish"); } yylval = yytext; return WORD_TOKEN; }  /* Handle individual words */
```
NIT: Error code should be `ERRCODE_FEATURE_NOT_SUPPORTED`.

Commit:
Merge remote-tracking branch 'upstream/BABEL_5_X_DEV' into BABEL-4382-5_X

Commit:
Merge remote-tracking branch 'upstream/BABEL_5_X_DEV' into BABEL-4382-5_X

Commit:
Support boolean operators grammar in contains search condition

Previously in Babelfish the support of Boolean operators was not enabled. To fix this we have created a grammar to allow customers to create search expressions with boolean operators 'AND', 'OR' and 'AND NOT' along with complex expressions using parenthesis over the simple_term and prefix_term operands.

TASK: BABEL-4382
Signed-off-by: Jaspal Singh ijaspals@amazon.com

Commit:
Support boolean operators grammar in contains search condition

Previously in Babelfish the support of Boolean operators was not enabled. To fix this we have created a grammar to allow customers to create search expressions with boolean operators 'AND', 'OR' and 'AND NOT' along with complex expressions using parenthesis over the simple_term and prefix_term operands.

TASK: BABEL-4382
Signed-off-by: Jaspal Singh ijaspals@amazon.com

Commit:
Support boolean operators grammar in contains search condition

Previously in Babelfish the support of Boolean operators was not enabled. To fix this we have created a grammar to allow customers to create search expressions with boolean operators 'AND', 'OR' and 'AND NOT' along with complex expressions using parenthesis over the simple_term and prefix_term operands.

TASK: BABEL-4382
Signed-off-by: Jaspal Singh ijaspals@amazon.com
